### ISSUE TITLE
[Bug]: `@babel/register` fails when `require()` a PURE esm module

### ISSUE THREAD

**Issue by n1ru4l** [babel/babel#14579]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

@babel/register or @babel/node

### Input code

**./test-subject/index.js**

```js
export default 'hi';

```

**./test-subject/package.json**

```js
{
  "name": "foo",
  "exports": {
    ".": "./index.js"
  },
  "type": "module"
}
```

**./index.js**
```js
require('@babel/register')({
  babelrc: false,
  cache: false,
  plugins: ['@babel/plugin-transform-modules-commonjs'],
});

require('./test-subject');
```

**./package.json**

```js
{
  "name": "node-starter",
  "version": "0.0.0",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "@babel/core": "^7.18.0",
    "@babel/plugin-transform-modules-commonjs": "^7.18.0",
    "@babel/register": "^7.17.7"
  }
}
```

Full runnable example over here:
https://stackblitz.com/edit/node-hwyjyp?file=index.js

### Configuration

NONE (see code)

### Current and expected behavior

The current behavior is that the following error is raised when running the script:

```
Error [ERR_REQUIRE_ESM]: require() of ES Module /home/projects/node-hwyjyp/test-subject/index.js from /home/projects/node-hwyjyp/index.js not supported.
Instead change the require of /home/projects/node-hwyjyp/test-subject/index.js in /home/projects/node-hwyjyp/index.js to a dynamic import() which is available in all CommonJS modules.
```

Instead, I expect babel to wrap the original `require` and skip the "ERR_REQUIRE_ESM" error check.

### Environment

```
    "@babel/core": "^7.18.0",
    "@babel/plugin-transform-modules-commonjs": "^7.18.0",
    "@babel/register": "^7.17.7",
```

Node version: v16.14.2

### Possible solution

A manual fix is to edit the package.json of the module, that, however, seems like a hack to me... Especially when you are dealing with an npm package that you have no control over (e.g. graphql@17)

**Comment by goodeath** [comment]
Hey man, currently I'm dealing with this module headache. I'm not part of the babel team, but let me know if I helped somehow.
We can check this on this link [@babel/register](https://babeljs.io/docs/en/babel-register#experimental-babel-8-implementation)
> Note: @babel/register does not support compiling native Node.js ES modules on the fly, since currently there is no stable API for intercepting ES modules loading.

As most know, ES Modules can't be fully converted in CommonJS modules. We can import ES Modules using dynamic import:
```javascript
require('@babel/register')({
  babelrc: false,
  cache: false,
  plugins: ['@babel/plugin-transform-modules-commonjs'],
});
import('./test-subject'); // Note it
```

Maybe we'd hit in some async issues, but, that's not the point. The thing that you wanted was

> Instead, I expect babel to wrap the original require and skip the "ERR_REQUIRE_ESM" error check.

As far as I know, there's no code workaround beyond that I presented before to replace "require" with some other function that actually works. 

And I think that ERR_REQUIRE_ESM shouldn't be skipped, because is a legit error. You tried to require (synchronously) a module that can be asynchronous, and CommonJS doesn't accept TLA(Top-Level Await)

Maybe I'm all wrong, I'm still trying to figure out how to use external ES Modules and CommonJS together in my project.

**Comment by n1ru4l** [comment]
@goodeath In this scenario, the ESM module is transpiled to CommonJS (using `@babel/plugin-transform-modules-commonjs`) before being required, so technically it is a CommonJS module.

This is kind of a weird situation because the Node.js `require` function is not aware of this transpilation process that babel does on the fly. On the other hand, it also does not seem to me that it would be the scope of babel/register to hijack the Node.js require for applying some module resolution patching (and introduce "allow require ESM module that is transpiled to CommonJS" behavior).

For now my solution was to build a script that runs as a post-install script which transpiles and edits the node_module. [You can find it here](https://github.com/graphql/graphql-js/issues/3603#issue-1246255464). A benefit of this approach is that you also do not need `@babel/register` anymore 🤔

**Comment by goodeath** [comment]
@n1ru4l I looked into your solution. Yeah, it works. You can do things like this, but I think there is one loophole: If some ESM module that you rewrited to use "require", use "await" outside of a function, it'd result in error. What you think? 

BTW, thanks for sharing your solution. Maybe it can come in handy sometime. I usually say that, one solution, even if it's not the better, is better than none.

**Comment by mbtts** [comment]
> it also does not seem to me that it would be (beyond) the scope of babel/register to hijack the Node.js require for applying some module resolution patching (and introduce "allow require ESM module that is transpiled to CommonJS" behavior).

I agree it is fair suggestion considering as Babel CLI has no problem transpiling the same sources.

If anyone coming late to this thread, like me, I have found these useful for helping to migrate to ESM:

* [@esbuild-kit/cjs-loader](https://github.com/esbuild-kit/cjs-loader): Loads ESM modules in a common CJS environment (using esbuild transform API)
    * Using [loaders](https://github.com/nodejs/loaders), which as mentioned above are still experimental
* [babel-plugin-module-extension-resolver](https://www.npmjs.com/package/babel-plugin-module-extension-resolver) Fixes imports to use `.js` extension
